### PR TITLE
add csv writer

### LINKED ISSUE #1237: CSV writer
A community user asked:

>  With some non-Zeek NDJSON data sources we turn into ZNG and store, we also may want to write to a SQL db (JSON works but prefer CSV) to populate visualizations of the data in say like an Apache Superset (which I actually have plans to integrate for Zeek data, probably back out in Zeek JSON from ZNG).

### PR DESCRIPTION
**Pull request by mccanne**
Fixes #1237

### CODE REVIEW

File: zio/csvio/writer.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	first      *zng.TypeRecord
+}
+
+func NewWriter(w io.WriteCloser, utf8, epochDates bool) *Writer {
+	var format zng.OutFmt
+	if utf8 {
+		format = zng.OutFormatZeek
+	} else {
+		format = zng.OutFormatZeekAscii
+	}
```
Comment (nwt) [inline zio/csvio/writer.go:32]:
```suggestion
	format := zng.OutFormatZeekAscii
	if utf8 {
		format = zng.OutFormatZeek
	}
```

File: zio/csvio/writer.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		return ErrNotDataFrame
+	}
+	var out []string
+	for k, col := range rec.Type.Columns {
+		var v string
+		value := rec.Value(k)
+		if !w.epochDates && col.Name == "ts" && col.Type == zng.TypeTime {
+			if value.IsUnsetOrNil() {
+				v = ""
+			} else {
```
Comment (nwt) [inline zio/csvio/writer.go:71]:
```suggestion
			if !value.IsUnsetOrNil() {
```

File: zio/csvio/writer.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+				ts, err := zng.DecodeTime(value.Bytes)
+				if err != nil {
+					return err
+				}
+				v = nano.Ts(ts).Time().UTC().Format(time.RFC3339Nano)
```
Comment (nwt) [inline zio/csvio/writer.go:76]:
```suggestion
				v = ts.Time().UTC().Format(time.RFC3339Nano)
```

File: zio/flags.go [hunk 4]
```diff
@@ -77,7 +77,7 @@ func (w *WriterFlags) SetFlags(fs *flag.FlagSet) {
 	fs.BoolVar(&w.UTF8, "U", false, "display zeek strings as UTF-8")
 	fs.BoolVar(&w.Text.ShowTypes, "T", false, "display field types in text output")
 	fs.BoolVar(&w.Text.ShowFields, "F", false, "display field names in text output")
-	fs.BoolVar(&w.Text.EpochDates, "E", false, "display epoch timestamps in text output")
+	fs.BoolVar(&w.EpochDates, "E", false, "display epoch timestamps in text output")
```
Comment (nwt) [inline zio/flags.go:80]:
```suggestion
	fs.BoolVar(&w.EpochDates, "E", false, "display epoch timestamps in csv and text output")
```

File: tests/formats/csv/hello.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        0:[goodnight;gracie;]
+        0:[,;gracie;]
+
+output-format: csv
+
+output: |
+        a,b
+        hello,world
+        goodnight,gracie
+        ",",gracie
```
Comment (nwt) [inline tests/formats/csv/hello.yaml:15]: Elsewhere we indent two spaces per level in YAML.
```suggestion
  #0:record[a:string,b:string]
  0:[hello;world;]
  0:[goodnight;gracie;]
  0:[,;gracie;]

output-format: csv

output: |
  a,b
  hello,world
  goodnight,gracie
  ",",gracie
```

File: tests/formats/csv/nested.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+input: |
+        #0:record[a:string,b:record[c:string,d:string]]
+        0:[hello;[world;goodbye;]]
+
+output-format: csv
+
+output: |
+        a,b.c,b.d
+        hello,world,goodbye
```
Comment (nwt) [inline tests/formats/csv/nested.yaml:11]: Elsewhere we indent two spaces per level in YAML.
```suggestion
  #0:record[a:string,b:record[c:string,d:string]]
  0:[hello;[world;goodbye;]]

output-format: csv

output: |
  a,b.c,b.d
  hello,world,goodbye
```

**Review by nwt — APPROVED** [review]

File: zio/csvio/writer.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	"io"
+	"time"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/flattener"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+var ErrNotDataFrame = errors.New("Cannot format non-dataframe group of records as csv.")
```
Comment (alfred-landrum) [inline zio/csvio/writer.go:15]:
```suggestion
var ErrNotDataFrame = errors.New("different record types seen during csv write")
```

File: zio/ndjsonio/config_test.go [hunk 8]
```diff
@@ -1,9 +1,10 @@
-package ndjsonio
+package ndjsonio_test
```
Comment (alfred-landrum) [inline zio/ndjsonio/config_test.go:1]: these ndjson changes looks unrelated to csv, are they needed?

Comment (mccanne) [reply zio/ndjsonio/config_test.go:1]: sorry, rebasing troubles.  will fix.

File: zio/csvio/writer.go (hunk 3 above)
Comment (mccanne) [reply zio/csvio/writer.go:76]: This came fromo textio.  Fixed there too.

File: zio/csvio/writer.go (hunk 7 above)
Comment (mccanne) [reply zio/csvio/writer.go:15]: Ha!  My first error message was like this one but I changed it to be more data-sciency.  Will change it back.

File: zio/csvio/writer.go (hunk 1 above)
Comment (mccanne) [reply zio/csvio/writer.go:32]: This came from textio and tableio.  Fixed there too.  Also, I needed to change const declaration in zng.go to cast to zng.OutFmt, which was needed anyway.
